hazelcast/src/main/java/com/hazelcast/sql/impl/exec/scan/MapScanExec.java

```diff
@@ -0,0 +1,130 @@
+import com.hazelcast.sql.impl.exec.IterationResult;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.extract.QueryTargetDescriptor;
+import com.hazelcast.sql.impl.row.HeapRow;
+import com.hazelcast.sql.impl.row.ListRowBatch;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.row.RowBatch;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.worker.QueryFragmentContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Executor for map scan.
+ */
+public class MapScanExec extends AbstractMapScanExec {
+    /** Batch size. To be moved outside when the memory management is ready. */
+    static final int BATCH_SIZE = 1024;
+
+    private final MapContainer map;
+    private final PartitionIdSet partitions;
+
+    private int migrationStamp;
+    private MapScanExecIterator recordIterator;
+
+    private List<Row> currentRows;
+
+    public MapScanExec(
+        int id,
+        MapContainer map,
+        PartitionIdSet partitions,
+        QueryTargetDescriptor keyDescriptor,
+        QueryTargetDescriptor valueDescriptor,
+        List<String> fieldNames,
+        List<QueryDataType> fieldTypes,
+        List<Integer> projects,
+        Expression<Boolean> filter,
+        InternalSerializationService serializationService
+    ) {
+        super(id, map.getName(), keyDescriptor, valueDescriptor, fieldNames, fieldTypes, projects, filter, serializationService);
+
+        this.map = map;
+        this.partitions = partitions;
+    }
+
+    @Override
+    protected void setup1(QueryFragmentContext ctx) {
+        migrationStamp = map.getMapServiceContext().getService().getMigrationStamp();
+        recordIterator = MapScanExecUtils.createIterator(map, partitions);
+    }
+
+    @Override
+    public IterationResult advance0() {
+        currentRows = null;
+
+        while (recordIterator.tryAdvance()) {
+            HeapRow row = prepareRow(recordIterator.getKey(), recordIterator.getValue());
+
+            if (row != null) {
+                if (currentRows == null) {
+                    currentRows = new ArrayList<>(BATCH_SIZE);
+                }
+
+                currentRows.add(row);
+
+                if (currentRows.size() == BATCH_SIZE) {
+                    break;
+                }
+            }
+        }
+
+        // Check for concurrent migration
+        if (!map.getMapServiceContext().getService().validateMigrationStamp(migrationStamp)) {
+            throw QueryException.error(SqlErrorCode.PARTITION_MIGRATED, "Map scan failed due to concurrent partition migration "
+                + "(result consistency cannot be guaranteed)");
+        }
+
+        // Check for concurrent map destroy
+        if (map.isDestoyed()) {
+            throw QueryException.error(SqlErrorCode.MAP_DESTROYED, "IMap has been destroyed concurrently: " + mapName);
+        }
+
+        boolean done = !recordIterator.hasNext();
```

Comment:
Minor: it would be "cleaner" to do the `hasNext` check before checking for migration stamp and destruction. It's not clear from the contract does the `hasNext` interact with the underlying store or not (which potentially can be destroyed or migrated).

Reply:
Fixed.